.github/workflows/ci.yml

```diff
@@ -1,180 +1,233 @@
-    name: Docker Alpine (Binary)
-    uses: stacks-network/stacks-blockchain/.github/workflows/image-build-alpine-binary.yml@master
-    needs: github-release
-    with:
-      tag: ${{ inputs.tag }}
-      docker_platforms: linux/arm64, linux/amd64, linux/amd64/v2, linux/amd64/v3
-    secrets:
-      DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
-      DOCKERHUB_PASSWORD: ${{ secrets.DOCKERHUB_PASSWORD }}
-
-  ## Create docker debian images
-  ## Only run if:
-  ##    - Tag is provided
-  ##    - OR
-  ##        - Not the default branch
-  ##        - AND
-  ##        - Not a PR
-  docker-debian:
-    if: ${{ inputs.tag != '' || (github.ref != format('refs/heads/{0}', github.event.repository.default_branch) && !contains(github.ref, 'refs/pull')) }}
-    name: Docker Debian (Binary)
-    uses: stacks-network/stacks-blockchain/.github/workflows/image-build-debian-binary.yml@master
-    needs: github-release
-    with:
-      tag: ${{ inputs.tag }}
-      docker_platforms: linux/amd64, linux/amd64/v2, linux/amd64/v3
-      linux_version: debian
-      build_type: binary
-    secrets:
-      DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
-      DOCKERHUB_PASSWORD: ${{ secrets.DOCKERHUB_PASSWORD }}
-
-  ###############################################
-  ## Build Branch/PR
-  ###############################################
-  ## Create docker debian images
-  ## Only run if:
-  ##   - Tag is *not* provided
-  build-branch:
-    if: ${{ inputs.tag == '' }}
-    name: Docker Debian (Source)
+    secrets: inherit
+
+  ## Build and push Debian image built from source
+  ##
+  ## Runs when:
+  ##   - tag is not provided
+  ##   and the following are not true:
+  ##     - PR review submitted (not approved)
+  ##     and any of:
+  ##       - PR review comment
+  ##       - PR change is requested
+  docker-image:
+    if: |
+      inputs.tag == '' &&
+      !(
+        github.event_name == 'pull_request_review' &&
+        github.event.action == 'submitted' &&
+        (
+          github.event.review.state == 'commented' ||
+          github.event.review.state == 'changes_requested'
+        )
+      )
+    name: Docker Image (Source)
     uses: stacks-network/stacks-blockchain/.github/workflows/image-build-debian-source.yml@master
     needs:
       - rustfmt
-      - leaked-cred-test
-    with:
-      docker_platforms: linux/amd64
-      linux_version: debian
-      build_type: source
-    secrets:
-      DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
-      DOCKERHUB_PASSWORD: ${{ secrets.DOCKERHUB_PASSWORD }}
+    secrets: inherit
+
+  ## Create a reusable cache for tests
+  ##
+  ## Runs when:
+  ##   - tag is provided
+  ##   or:
+  ##     - no tag provided
+  ##       and any of:
+  ##         - PR is approved (any approval will trigger)
+  ##         - this workflow is called manually
+  ##         - PR is opened
+  ##         - commit to either (development, master) branch
+  create-cache:
```

Comment:
I noticed the caches being created (save for the bitcoin cache) is using the commit sha as part of the restore key, which as you know is pretty standard. However, if the cache is only being created at the end of a PR's life, when it gets approved, I wouldn't imagine it would get used very much. Perhaps only if there are more than 1 approval.

On a similar note, why have anything trigger on approvals? Maybe there are some critical workflows that must only run after approvals which I'm missing, but my initial thoughts are this wouldn't be needed as long as the right workflows run every time a PR is synchronized.

Reply:
the cache is actually created anytime the `ci.yml` (or any of the test workflows, if triggered manually) is run. 
so, on PR creation, synchronization (new commit), or a PR approval - after rustfmt, the next thing that happens is the cache is either confirmed to exist or it is created for the tests that need to be run. 

Ideally, at the end of PR's approval - the cache may still exist so that action will save ~15 minutes of runtime since the commit sha will match the key (as you noted) of  an existing cache (this is also dependent on the timing of the `cleanup.yml` workflow, which is a bit aggressive currently). 

the reason for triggering workflows on PR approval is simple - i'm proposing that for opening a PR, PR synchronization, we run a subset of tests (bitcoin tests, unit tests) vs a PR approval which would run *all* tests we require for a release.  The reasoning behind this was simply to save time on running the same tests multiple time over the course of a PR.  There is an argument to be made that perhaps the only tests that should not be run as part of a PR action is what's defined in `slow-tests.yml`, this was simply the choice I've made here. 

I felt that running a limited number of tests for opening/synchronizing a PR was a good compromise as opposed to running all the tests based on any PR change (ex: if you forget to add a changelog to your PR - the workflow still needs to run currently, even though no code has changed). 

the second compromise was adding a manual way to run the tests in lieu of running them all every action run (`standalone-tests.yml`)

Reply:
Makes sense to me! I missed that caches were being created on every run, thanks for clarifying.